kube-runtime/src/utils/stream_backoff.rs

```diff
@@ -0,0 +1,189 @@
+                        duration = ?backoff_duration,
+                        "Error received, backing off"
+                    );
+                    this.state.set(State::BackingOff(backoff_sleep));
+                } else {
+                    tracing::debug!("Error received, giving up");
+                    this.state.set(State::GivenUp);
+                }
+            }
+            Poll::Ready(_) => {
+                tracing::trace!("Non-error received, resetting backoff");
+                this.backoff.reset();
+            }
+            Poll::Pending => {}
+        }
+        next_item
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{task::Poll, time::Duration};
+
+    use super::StreamBackoff;
+    use backoff::backoff::Backoff;
+    use futures::{channel::mpsc, pin_mut, poll, stream, StreamExt};
+
+    #[tokio::test]
+    async fn stream_should_back_off() {
+        tokio::time::pause();
+        let tick = Duration::from_secs(1);
+        let rx = stream::iter([Ok(0), Ok(1), Err(2), Ok(3), Ok(4)]);
+        let rx = StreamBackoff::new(rx, backoff::backoff::Constant::new(tick));
+        pin_mut!(rx);
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(0))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(1))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(2))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(tick * 2).await;
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(3))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(4))));
+        assert_eq!(poll!(rx.next()), Poll::Ready(None));
+    }
+
+    #[tokio::test]
+    async fn backoff_time_should_update() {
+        tokio::time::pause();
+        let (tx, rx) = mpsc::unbounded();
+        // let rx = stream::iter([Ok(0), Ok(1), Err(2), Ok(3)]);
+        let rx = StreamBackoff::new(rx, LinearBackoff::new(Duration::from_secs(2)));
+        pin_mut!(rx);
+        tx.unbounded_send(Ok(0)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(0))));
+        tx.unbounded_send(Ok(1)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(1))));
+        tx.unbounded_send(Err(2)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(2))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(3)).await;
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tx.unbounded_send(Err(3)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Err(3))));
+        tx.unbounded_send(Ok(4)).unwrap();
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(3)).await;
+        assert_eq!(poll!(rx.next()), Poll::Pending);
+        tokio::time::advance(Duration::from_secs(2)).await;
+        assert_eq!(poll!(rx.next()), Poll::Ready(Some(Ok(4))));
+        assert_eq!(poll!(rx.next()), Poll::Pending);
```

Comment:
Maybe add a // pending because stream empty or // pending because backing off N seconds
on the `assert_eq!(poll!(rx.next()), Poll::Pending);` lines for readability.

Reply:
i can add that.